### ISSUE TITLE
追加学習

### ISSUE THREAD

**Issue by yonedahayato** [yonedahayato/re_training_open_clip#1]
- [x] データの確認
- [x] ロスの確認
- [x] 追加学習の実行

**Comment by yonedahayato** [comment]
# データセットの確認

In addition to specifying the training data via CSV files as mentioned above, our codebase also supports [webdataset](https://github.com/webdataset/webdataset), which is recommended for larger scale datasets. The expected format is a series of .tar files. Each of these .tar files should contain two files for each training example, one for the image and one for the corresponding text. Both files should have the same name but different extensions. For instance, shard_001.tar could contain files such as abc.jpg and abc.txt. You can learn more about webdataset at https://github.com/webdataset/webdataset. We use .tar files with 1,000 data points each, which we create using [tarp](https://github.com/webdataset/tarp).

You can download the YFCC dataset from [Multimedia Commons](http://mmcommons.org/). Similar to OpenAI, we used a subset of YFCC to reach the aforementioned accuracy numbers. The indices of images in this subset are in [OpenAI's CLIP repository](https://github.com/openai/CLIP/blob/main/data/yfcc100m.md).

> webdataset を利用

---

上記のようにCSVファイルを介してトレーニングデータを指定することに加えて、コードベースはWebデータセットもサポートしています。これは大規模なデータセットに推奨されます。 予想される形式は、一連の.tarファイルです。 これらの各.tarファイルには、トレーニング例ごとに2つのファイルが含まれている必要があります。1つは画像用で、もう1つは対応するテキスト用です。 両方のファイルの名前は同じですが、拡張子は異なります。 たとえば、shard_001.tarには、abc.jpgやabc.txtなどのファイルを含めることができます。 webdatasetの詳細については、https：//github.com/webdataset/webdatasetをご覧ください。 tarpを使用して作成した1,000個のデータポイントを持つ.tarファイルを使用します。

YFCCデータセットはMultimediaCommonsからダウンロードできます。 OpenAIと同様に、YFCCのサブセットを使用して、前述の精度の数値に到達しました。 このサブセットの画像のインデックスは、OpenAIのCLIPリポジトリにあります。

**Comment by yonedahayato** [comment]
- [x] get_wds_dataset の挙動の確認

---

https://github.com/yonedahayato/re_training_open_clip/blob/ef91b858bbf1cad9984d6947605a4a97dd90f0b1/src/training/data.py#L188-L281

---

https://github.com/yonedahayato/re_training_open_clip/blob/ef91b858bbf1cad9984d6947605a4a97dd90f0b1/src/training/main.py#L253

---

https://github.com/yonedahayato/re_training_open_clip/blob/ef91b858bbf1cad9984d6947605a4a97dd90f0b1/src/training/train.py#L47-L143

---

- dataloader 取り出し

https://github.com/yonedahayato/re_training_open_clip/blob/ef91b858bbf1cad9984d6947605a4a97dd90f0b1/src/training/train.py#L60

---

- for で iteration

https://github.com/yonedahayato/re_training_open_clip/blob/ef91b858bbf1cad9984d6947605a4a97dd90f0b1/src/training/train.py#L70

---

- batch からの取り出し

https://github.com/yonedahayato/re_training_open_clip/blob/ef91b858bbf1cad9984d6947605a4a97dd90f0b1/src/training/train.py#L74

---

- train_data の設置

```
--train-data '/data/cc12m/cc12m-train-{0000..2175}.tar'
```

**Comment by yonedahayato** [comment]
# Loss の確認

- Loss の計算

https://github.com/yonedahayato/re_training_open_clip/blob/ef91b858bbf1cad9984d6947605a4a97dd90f0b1/src/training/train.py#L81-L83

---

- Loss の作成

https://github.com/yonedahayato/re_training_open_clip/blob/ef91b858bbf1cad9984d6947605a4a97dd90f0b1/src/training/train.py#L51-L58

--- 

- Loss の設計

https://github.com/yonedahayato/re_training_open_clip/blob/ef91b858bbf1cad9984d6947605a4a97dd90f0b1/src/open_clip/loss.py#L58-L114

**Comment by yonedahayato** [comment]
# 追加学習の実行

```
データの確認 cherry blossoms and a field of roses make for a classic feminine scene celebrating perfume .
train loss:  tensor(3.6071, grad_fn=<DivBackward0>)
test loss:  tensor(2.8353)
データの確認 cherry blossoms and a field of roses make for a classic feminine scene celebrating perfume .
train loss:  tensor(3.4811, grad_fn=<DivBackward0>)
test loss:  tensor(2.7836)
データの確認 cherry blossoms and a field of roses make for a classic feminine scene celebrating perfume .
train loss:  tensor(3.0681, grad_fn=<DivBackward0>)
test loss:  tensor(2.7679)
データの確認 cherry blossoms and a field of roses make for a classic feminine scene celebrating perfume .
train loss:  tensor(2.7798, grad_fn=<DivBackward0>)
test loss:  tensor(2.7503)
データの確認 cherry blossoms and a field of roses make for a classic feminine scene celebrating perfume .
train loss:  tensor(2.2913, grad_fn=<DivBackward0>)
test loss:  tensor(2.7236)
```
